url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/zmq/Trie.java

```diff
@@ -0,0 +1,298 @@
+
+
+public class Trie {
+    private int refcnt;
+
+    private byte min;
+    private short count;
+    private short live_nodes;
+    
+    public static interface ITrieHandler {
+        void added(byte[] data, int size, Object arg);
+    }
+    Trie[] next;
+    
+    public Trie() {
+        min = 0;
+        count = 0;
+        live_nodes = 0;
+        
+        refcnt = 0;
+        next = null;
+    }
+    
+    //  Add key to the trie. Returns true if this is a new item in the trie
+    //  rather than a duplicate.
+    public boolean add (byte[] prefix_)
+    {
+        return add (prefix_, 0);
+    }
+    
+    public boolean add (byte[] prefix_, int start_)
+    {
+        //  We are at the node corresponding to the prefix. We are done.
+        if (prefix_ == null || prefix_.length == start_) {
+            ++refcnt;
+            return refcnt == 1;
+        }
+
+        byte c = prefix_[start_];
+        if (c < min || c >= min + count) {
+
+            //  The character is out of range of currently handled
+            //  charcters. We have to extend the table.
+            if (count == 0) {
+                min = c;
+                count = 1;
+                next = null;
+            }
+            else if (count == 1) {
+                byte oldc = min;
+                Trie oldp = next[0];
+                count = (short) ((min < c ? c - min : min - c) + 1);
+                next = new Trie[count];
+                min = (byte)Math.min ((byte)min, (byte)c);
+                next[oldc - min] = oldp;
+            }
+            else if (min < c) {
+
+                //  The new character is above the current character range.
+                count = (short) (c - min + 1);
+                next = realloc(next, count, true);
+            }
+            else {
+
+                //  The new character is below the current character range.
+                count = (short) ((min + count) - c);
+                next = realloc(next, count, false);
+                min = c;
+            }
+        }
+
+        //  If next node does not exist, create one.
+        if (count == 1) {
+            if (next == null) {
+                next = new Trie[1];
+                next[0] = new Trie();
+                ++live_nodes;
+                //alloc_assert (next.node);
+            }
+            return next[0].add (prefix_, start_ + 1);
+        }
+        else {
+            if (next[c - min] == null) {
+                next[c - min] = new Trie();
+                ++live_nodes;
+                //alloc_assert (next.table [c - min]);
+            }
+            return next[c - min].add (prefix_ , start_ + 1);
+        }
+    }
+    
+    private Trie[] realloc(Trie[] table, short size, boolean ended) {
+        return Utils.realloc(Trie.class, table, size, ended);
+    }
+    
+    //  Remove key from the trie. Returns true if the item is actually
+    //  removed from the trie.
+    public boolean rm (byte[] prefix_, int start_)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 20. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 123-128 into a separate method.